### PR TITLE
Error handling recipe - First feedback

### PR DESCRIPTION
**Pull request by goldbergyoni**
I'm landing here a draft code of the error handling testing. This recipe will show how important and rich are error-handling tests - Many scenarios and error types could occur

Thoughts before I accomplish this?

image

### CODE REVIEW

File: example-application/api-under-test.js [hunk 1]
```diff
@@ -82,20 +81,22 @@ const defineRoutes = (expressApp) => {
...
     }
+    await errorHandler.handleError(err);
     res.status(500).end();
   });
 };
 
-process.on("uncaughtException", () => {
-  // a log of other logic here
-  console.log("Error occured!");
+process.on("uncaughtException", (error) => {
```
Comment (mikicho) [inline example-application/api-under-test.js:94]: NIT: jsut for the uniformity
```suggestion
process.on("uncaughtException", (err) => {
```

File: example-application/api-under-test.js [hunk 2]
```diff
@@ -82,20 +81,22 @@ const defineRoutes = (expressApp) => {
...
+  errorHandler.handleError(error);
+});
+
+process.on("unhandledRejection", (reason, originPromise) => {
```
Comment (mikicho) [inline example-application/api-under-test.js:98]:
```suggestion
process.on("unhandledRejection", (reason) => {
```

File: example-application/data-access/config/config.js [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-  username: 'myuser',
-  password: 'myuserpassword',
-  database: 'shop',
-  host: 'localhost',
-  port: 54320,
+  username: "myuser",
+  password: "myuserpassword",
+  database: "shop",
+  host: "localhost",
+  port: 54310,
```
Comment (mikicho) [inline example-application/data-access/config/config.js:6]: why?

File: example-application/error-handling.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const mailer = require("./libraries/mailer");
+
+// This file simulates real-world error handler that makes this component observable
+const errorHandler = {
+  handleError: async (errorToHandle) => {
+    console.error(errorToHandle);
+    metricsExporter.fireMetric("error", { errorName: errorToHandle.name || "generic-error" });
```
Comment (mikicho) [inline example-application/error-handling.js:7]: 🔥

File: example-application/api-under-test.js [hunk 5]
```diff
@@ -82,20 +81,22 @@ const defineRoutes = (expressApp) => {
...
-    if (process.env.SEND_MAILS === "true") {
-      // important notification logic here
-      await mailer.send("Error", err.message, "admin@app.com");
-
-      // Other important notification logic here
+    if (typeof err === "object") {
+      if (err.isTrusted === undefined || err.isTrusted === null) {
+        err.isTrusted = true; //Error during a specific request is usually not catastrophic and should not lead to process exit
+      }
     }
```
Comment (mikicho) [inline example-application/api-under-test.js:88]:
```suggestion
      err?.isTrusted = true; //Error during a specific request is usually not catastrophic and should not lead to process exit
```

File: example-application/test/basic-tests.test.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ const request = require("supertest");
 const sinon = require("sinon");
 const nock = require("nock");
 const { initializeWebServer, stopWebServer } = require("../api-under-test");
-const mailer = require("../libraries/mailer");
+const mailer = require("../error-handling").mailer;
```
Comment (mikicho) [inline example-application/test/basic-tests.test.js:5]: 1. how is this working? you don't export `mailer` in the `error-handling.js` file
2. This is weird that we need to import the mailer from `error-handling` file, what's wrong with what we have?

File: various-receipes/error-handling/test/error-handling.tests.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+beforeEach(() => {
+  nock.cleanAll();
+  // ️️️✅ Best Practice: Isolate the service under test by intercepting requests to 3rd party services
+  nock("http://localhost/user/").get(`/1`).reply(200, {
+    id: 1,
+    name: "John",
+  });
+  nock("https://mailer.com").post("/send").reply(202);
+  sinon.restore();
```
Comment (mikicho) [inline various-receipes/error-handling/test/error-handling.tests.js:35]: Move to `afterEach`? (also the `nock.cleanAll`)

File: various-receipes/error-handling/test/error-handling.tests.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  describe("Various Throwing Scenarios And Locations", () => {
+    test.todo("When an error is thrown during startup, then it's handled correctly");
+    test.todo("When an error is thrown during web request, then it's handled correctly");
+    test.todo("When an error is thrown during a queue message processing , then it's handled correctly");
+    test.todo("When an error is thrown from a timer, then it's handled correctly");
+    test.todo("When an error is thrown from a middleware, then it's handled correctly");
+  });
+
+  describe("Various Error Types", () => {
+    test.each`
```
Comment (mikicho) [inline various-receipes/error-handling/test/error-handling.tests.js:111]: wasn't familiar with this table-like syntax. very cool!

File: example-application/error-handling.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    // In real world - The right flow is sending alerts from the monitoring system
+    mailer.send("Error occured", errorToHandle, "admin@our-domain.io");
+
+    // A common best practice is to crash when an unknown error (non-trusted) is being thrown
+    decideWhetherToCrash(errorToHandle);
+  },
+};
+
+const decideWhetherToCrash = (error) => {
+  if (!error.isTrusted) {
```
Comment (mikicho) [inline example-application/error-handling.js:18]: If I understand the code correctly, this will crash our app:

Comment (mikicho) [inline example-application/error-handling.js:18]: If I understand the code correctly, this will crash our app: `throw 1;`

**Review by mikicho — CHANGES_REQUESTED** [review]
I learned so much from this PR. thanks!

File: example-application/api-under-test.js (hunk 1 above)
Comment (goldbergyoni) [reply example-application/api-under-test.js:94]: Uniformity with what? I usually prefer to use the entire word:

https://github.com/kettanaito/naming-cheatsheet#avoid-contractions

File: example-application/data-access/config/config.js (hunk 3 above)
Comment (goldbergyoni) [reply example-application/data-access/config/config.js:6]: Another project in my computer is already using 54320/5432

File: example-application/api-under-test.js (hunk 5 above)
Comment (goldbergyoni) [reply example-application/api-under-test.js:88]: Can I do optional chaining with an assignment? 🤔

File: example-application/error-handling.js (hunk 9 above)
Comment (goldbergyoni) [reply example-application/error-handling.js:18]: @mikicho Yes, unless you throw 1 from a inside a request handler. If you throw 1 on process start - Have to assume that something unknown and terrible happened

File: various-receipes/error-handling/test/error-handling.tests.js (hunk 7 above)
Comment (goldbergyoni) [reply various-receipes/error-handling/test/error-handling.tests.js:35]: Done

File: example-application/test/basic-tests.test.js (hunk 6 above)
Comment (goldbergyoni) [reply example-application/test/basic-tests.test.js:5]: It was a mistake, we don't even use this, Fixed.

File: example-application/api-under-test.js (hunk 1 above)
Comment (mikicho) [reply example-application/api-under-test.js:94]: With line `83` for example.
but, it's NIT.
and I'm ok with both `err` and `error`, we just choose one of them

File: example-application/api-under-test.js (hunk 5 above)
Comment (mikicho) [reply example-application/api-under-test.js:88]: you're right, my mistake..
I just toy with js a lil bit and you can also do:
```
typeof err === "object" && !err.isTrusted && (err.isTrusted = true)
```

But you can resolve it :D

Comment (mikicho) [reply example-application/api-under-test.js:88]: you're right, my mistake..

File: example-application/error-handling.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+const mailer = require("./libraries/mailer");
+
+// This file simulates real-world error handler that makes this component observable
+const errorHandler = {
+  handleError: async (errorToHandle) => {
+    console.error(errorToHandle);
+    metricsExporter.fireMetric("error", { errorName: errorToHandle.name || "generic-error" });
+    // This is used to simulate sending email to admin when an error occurs
+    // In real world - The right flow is sending alerts from the monitoring system
+    mailer.send("Error occured", errorToHandle, "admin@our-domain.io");
```
Comment (mikicho) [inline example-application/error-handling.js:10]:
```suggestion
    await mailer.send("Error occured", errorToHandle, "admin@our-domain.io");
```

File: example-application/error-handling.js (hunk 9 above)
Comment (mikicho) [reply example-application/error-handling.js:18]: > unless you throw 1 from a inside a request handler

@goldbergyoni are you sure?
we don't wrap errors from the routers..


```
 expressApp.use(async (err, req, res, next) => {
    if (typeof err === "object") {
      if (err.isTrusted === undefined || err.isTrusted === null) {
        err.isTrusted = true; //Error during a specific request is usually not catastrophic and should not lead to process exit
      }
    }
    await errorHandler.handleError(err);
    res.status(500).end();
  });
```

File: example-application/api-under-test.js (hunk 1 above)
Comment (goldbergyoni) [reply example-application/api-under-test.js:94]: Gotcha, chose error and fixed

File: example-application/error-handling.js (hunk 10 above)
Comment (goldbergyoni) [reply example-application/error-handling.js:10]: Done

File: example-application/error-handling.js (hunk 9 above)
Comment (goldbergyoni) [reply example-application/error-handling.js:18]: @mikicho I added a test for this flow now - Throwing unknown error during a request, the process stays alive. The post route has try-catch.

**Comment by goldbergyoni** [comment]
@mikicho Kindly re-review

**Pull request merged**